File: iOS实用技术/RunLoop/RunLoop实现常驻线程/RunLoop实现常驻线程/ViewController.m
//
//  ViewController.m
//  RunLoop实现常驻线程
//
//  Created by 陈登文 on 2021/8/25.
//

#import "ViewController.h"
#import "RunLoopObject.h"

@interface ViewController ()
{
    NSThread *_thread;
}
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    
    NSThread *thread = [RunLoopObject threadDispatch];
    NSLog(@"current thread == %p, name == %@", [NSThread currentThread], [NSThread currentThread].name);
    NSLog(@"thread == %p, name == %@", thread, thread.name);
    
//    sleep(2);
//    NSLog(@"thread == %p, name == %@", thread, thread.name);
}

- (IBAction)buttonClicked:(id)sender {
//    NSLog(@"thread == %p, name == %@", _thread, _thread.name);
}



@end

File: Swift/泊学付费网站Swift资料/（11）RxSwift/9.3 swift_alamofire/swift_alamofire/ViewController.swift
//
//  ViewController.swift
//  swift_alamofire
//
//  Created by gary on 2021/9/1.
//

import UIKit
import Alamofire
import RxSwift

class ViewController: UIViewController {
    
    var todoList = [Todo]()
    var bag = DisposeBag()
    
    @IBOutlet weak var tableView: UITableView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        tableView.dataSource = self
        
        let todoId:Int? = nil
        Observable.of(todoId)
        .map { tid in   // map 后转换为 Observable<TodoRouter>
            return TodoRouter.get(tid)
        }.flatMap { router in    // flatMap 会把原序列中的每一个事件，变成一个新的 Observable
            return Todo.getList(from: router)
        }.subscribe(onNext: { (todos:[[String : Any]]) in
            self.todoList = todos.compactMap{ Todo(json: $0) }
            self.tableView.reloadData()
        }, onError: { error in
            print(error.localizedDescription)
        }).disposed(by: bag)
        
        //传统习惯写法
//        Todo.getList(from: TodoRouter.get(nil))
//        .subscribe(onNext: { (todos:[[String : Any]]) in
//            self.todoList = todos.compactMap{ Todo(json: $0) }
//            self.tableView.reloadData()
//        }, onError: { error in
//            print(error.localizedDescription)
//        }).disposed(by: bag)
        
    }
}

extension ViewController: UITableViewDataSource {
    func tableView(_ tableView: UITableView,
                   numberOfRowsInSection section: Int) -> Int {
        return self.todoList.count
    }
    
    func tableView(_ tableView: UITableView,
                   cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(
            withIdentifier: "TodoItem", for: indexPath)
        let todo = todoList[indexPath.row]
        
        cell.textLabel?.text = todo.title
        
        return cell
    }
}

